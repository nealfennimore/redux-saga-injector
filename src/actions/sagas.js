import { PREFIX } from 'src/prefix';

export const RUN_SAGAS = `${PREFIX}/RUN_SAGAS`;
export function runSagas( sagas, uid ) {
    return {
        type: RUN_SAGAS,
        sagas,
        uid
    };
}

export const SAGAS_FINISHED = `${PREFIX}/SAGAS_FINISHED`;
export function sagasFinished( uid ) {
    return {
        type: SAGAS_FINISHED,
        uid
    };
}

export const CANCEL_SAGAS = `${PREFIX}/CANCEL_SAGAS`;
export function cancelSagas( uid ) {
    return {
        type: CANCEL_SAGAS,
        uid
    };
}

export const QUEUE_EMPTY = `${PREFIX}/QUEUE_EMPTY`;
export function queueEmpty() {
    return {
        type: QUEUE_EMPTY,
    };
}
