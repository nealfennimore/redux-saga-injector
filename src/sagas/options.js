/**
 * Default options
 * @export
 */
const defaultOptions = {
    preloadTimeout: 5,
    sagaTimeout: 5000
};

export default defaultOptions;
