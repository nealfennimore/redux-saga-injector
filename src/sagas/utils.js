/* eslint-disable import/prefer-default-export */
import { delay } from 'redux-saga';
import { call } from 'redux-saga/effects';

/**
 * Timeout
 * @export
 * @param {number} duration Duration of timeout
 */
export function* timeout( duration ) {
    return yield call( delay, duration );
}
