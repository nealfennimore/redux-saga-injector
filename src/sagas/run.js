import { delay, END } from 'redux-saga';
import { call, take, all, race, put, cancel, takeEvery } from 'redux-saga/effects';
import { TASK } from 'redux-saga/utils';
import * as actions from 'src/actions/sagas';
import { __BROWSER__ } from 'src/env';

/**
 * Determine if redux saga task
 *
 * @param {Object} task
 */
export function isTask( task ) {
    return !! ( typeof task === 'object' && task[TASK] );
}

/**
 * Cancel any saga that returns a task
 * Mostly intended for cancelling fork/spawn
 *
 * @export
 * @param {Array} sagas array of sagas
 */
export function* cancelTasks( sagas ) {
    const tasks = sagas.filter( isTask );
    yield cancel( ...tasks );
}

/**
 * Run all sagas by `call`
 *
 * @export
 * @param {Array} sagas Array of generator functions
 * @returns {Array}
 */
export function* runAll( sagas ) {
    return yield all(
        sagas.map( saga => call( saga ) )
    );
}

/**
 * Tell if an action is a valid cancel based on the uid
 *
 * @export
 * @param {String} uid
 * @param {Object} action
 * @returns {Boolean}
 */
export function isValidCancel( uid, action ) {
    return action.type === actions.CANCEL_SAGAS &&
    action.uid === uid;
}

/**
 * Watches until sagas based on uid are cancelled
 *
 * @export
 * @param {String} uid
 */
export function* watchCancellation( uid ) {
    while ( true ) {
        const action = yield take( actions.CANCEL_SAGAS );
        const isCancelled = yield call( isValidCancel, uid, action );
        if( isCancelled ) {
            break;
        }
    }
}

/**
 * Timeout function
 *
 * @export
 */
export function* timeout() {
    // Never timeout when browser
    if ( __BROWSER__ ) {
        return yield take( '__NEVER_END__' );
    }
    yield call( delay, 5000 );
    yield put( END );
}

/**
 * Run injected sagas
 *
 * @export
 * @param {Object} action
 */
export function* runSagas( action ) {
    try {
        const result = yield race( {
            sagas: call( runAll, action.sagas ),
            cancelled: call( watchCancellation, action.uid ),
            timeout: call( timeout ),
        } );
        if( result.sagas ) {
            yield put( actions.sagasFinished( action.uid ) );
        } else {
            yield call( cancelTasks, action.sagas );
            yield put( actions.cancelSagas( action.uid ) );
        }
    } catch ( error ) {
        yield put( actions.cancelSagas( action.uid ) );
    }
}


/**
 * Run sagas on the client
 *
 * @export
 */
export function* sagaRunner() {
    yield takeEvery( actions.RUN_SAGAS, runSagas );
}
