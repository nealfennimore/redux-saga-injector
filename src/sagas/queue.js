import { delay } from 'redux-saga';
import { call, take, put, spawn, cancel, takeEvery } from 'redux-saga/effects';
import * as actions from 'src/actions/sagas';
import { runSagas } from 'src/sagas/run';


/**
 * Adds a uid to the queue
 *
 * @export
 * @param {Set} queue
 * @param {Object} action
 */
export function addToQueue( queue, action ) {
    queue.add( action.uid );
}

/**
 * Removes a uid to the queue
 *
 * @export
 * @param {Set} queue
 * @param {Object} action
 */
export function removeFromQueue( queue, action ) {
    queue.delete( action.uid );
}

/**
 * Starts up running the sagas if there are any
 *
 * @export
 * @param {Set} queue
 * @param {Object} action
 */
export function* startup( queue, action ) {
    if( action.sagas.length ) {
        yield call( addToQueue, queue, action );
        yield spawn( runSagas, action );
    }
}

/**
 * Removes an item from the queue and dispatches when empty
 *
 * @export
 * @param {Set} queue
 * @param {Object} action
 */
export function* waitTilEmpty( queue, action ) {
    yield call( removeFromQueue, queue, action );
    if( ! queue.size ) {
        yield put( actions.queueEmpty() );
    }
}

/**
 * Creates a queue
 *
 * @export
 * @returns Set
 */
export function createQueue() {
    return new Set();
}

/**
 * Saga for server side rendering
 * Creates a queue, runs sagas, and completes when queue is empty
 *
 * @export
 */
export function* serverQueueSaga() {
    const queue = yield call( createQueue );
    const runTask = yield takeEvery( actions.RUN_SAGAS, startup, queue );
    const emptyTask = yield takeEvery( [actions.SAGAS_FINISHED, actions.CANCEL_SAGAS], waitTilEmpty, queue );

    // Give time to receive sagas
    yield call( delay, 0 );

    // Receive no more sagas
    yield cancel( runTask );

    // If we have a queue, then wait til all sagas finish or are cancelled
    if( queue.size ) {
        yield take( actions.QUEUE_EMPTY );
    }

    // Clean up remaining task
    yield cancel( emptyTask );

    // Done
}
