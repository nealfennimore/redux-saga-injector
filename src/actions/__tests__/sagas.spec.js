import * as actions from '../sagas';

describe( 'Saga actions', ()=>{
    test( 'runSagas', ()=>{
        expect(
            actions.runSagas( [jest.fn()], 'uid' )
        ).toMatchSnapshot();
    } );
    test( 'sagasFinished', ()=>{
        expect(
            actions.sagasFinished( 'uid' )
        ).toMatchSnapshot();
    } );
    test( 'cancelSagas', ()=>{
        expect(
            actions.cancelSagas( 'uid' )
        ).toMatchSnapshot();
    } );
    test( 'queueEmpty', ()=>{
        expect(
            actions.queueEmpty()
        ).toMatchSnapshot();
    } );
} );
