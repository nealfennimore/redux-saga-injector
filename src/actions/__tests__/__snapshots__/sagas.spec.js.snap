// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga actions cancelSagas 1`] = `
Object {
  "type": "@@redux-saga-injector/CANCEL_SAGAS",
  "uid": "uid",
}
`;

exports[`Saga actions queueEmpty 1`] = `
Object {
  "type": "@@redux-saga-injector/QUEUE_EMPTY",
}
`;

exports[`Saga actions runSagas 1`] = `
Object {
  "sagas": Array [
    [MockFunction],
  ],
  "type": "@@redux-saga-injector/RUN_SAGAS",
  "uid": "uid",
}
`;

exports[`Saga actions sagasFinished 1`] = `
Object {
  "type": "@@redux-saga-injector/SAGAS_FINISHED",
  "uid": "uid",
}
`;
